cmd/bnsd/x/statefix/handlers.go

```diff
@@ -0,0 +1,90 @@
+package statefix
+
+import (
+	weave "github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/migration"
+	"github.com/iov-one/weave/orm"
+	"github.com/iov-one/weave/x"
+)
+
+// RegisterRoutes registers handlers for message processing.
+func RegisterRoutes(r weave.Registry, auth x.Authenticator) {
+	r = migration.SchemaMigratingRegistry("statefix", r)
+	bucket := NewExecutedFixBucket()
+	r.Handle(&ExecuteFixMsg{}, &executeFixHandler{
+		auth:   auth,
+		bucket: bucket,
+	})
+}
+
+type executeFixHandler struct {
+	auth   x.Authenticator
+	bucket orm.ModelBucket
+}
+
+func (h *executeFixHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	if _, _, err := h.validate(ctx, db, tx); err != nil {
+		return nil, errors.Wrap(err, "invalid message")
+	}
+	return &weave.CheckResult{GasAllocated: 0}, nil
+}
+
+func (h *executeFixHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, fixFn, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid message")
+	}
+
+	if err := fixFn(ctx, db); err != nil {
+		return nil, errors.Wrap(err, "fix function failed")
+	}
+
+	// Create a record in the database to remember that this fix was executed.
+	fix := &ExecutedFix{
+		Metadata: &weave.Metadata{},
+		FixID:    msg.FixID,
+	}
+	if h.bucket.Put(db, []byte(msg.FixID), fix); err != nil {
+		return nil, errors.Wrap(err, "cannot persist information about executed fix")
+	}
+
+	return &weave.DeliverResult{}, nil
+}
+
+func (h *executeFixHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*ExecuteFixMsg, FixFunc, error) {
+	//
+	// TODO XXX
+	//
+	// If possible and makes sense, require signature of the whole
```

Comment:
I believe there must an authentication mechanism to prevent random actors to trigger a state fix. Just not sure if only `gov` sig is enough for this change.

Reply:
This is a design decision.

There is nothing harmful in anyone is allowed to do this. Code must be first deployed and it can be executed only once. In practice, you would like to execute the state fix as soon as it is deployed in enough places (majority).

I thought adding governance authentication might be a nice touch, that shows that any extraordinary change is approved by the governance as well (not only validators). I think this is for @davepuchyr to decide what is our decision on this.